File: lintcode/golang/0931_median_of_k_sorted_arrays.go
/**
 * @param nums: the given k sorted arrays
 * @return: the median of the given k sorted arrays
 */

import "math"

func findMedian (nums [][]int) float64 {
    // write your code here
    var n int
    for _, num := range nums {
        if len(num) == 0 || (len(num)== 1 && num[0] == 0) {
            continue
        }
        n += len(num)
    }
    
    if n % 2 == 1 {
        return findK(nums, n/2+1)
    }
    return (findK(nums, n/2) + findK(nums, n/2+1))/2
}

func findK (nums [][]int, target int) float64 {
    start, end := getRange(nums)
    
    for start + 1 < end {
        mid := start + (end-start)/2
        cnt := getCnt(nums, mid)
        if cnt >= target {
            end = mid
        } else {
            start = mid
        }
    }
    
    if getCnt(nums, start) >= target {
        return float64(start)
    }
    return float64(end)
}

func getCnt (nums [][]int, target int) int {
    var cnt int
    
    for _, num := range nums {
        cnt += getNolarger(num, target)
    }
    return cnt
}

func getNolarger(num []int, target int) int {
    n := len(num)
    if n == 0 || (n==1 && num[0] == 0) {
        return 0
    }
    
    l := 0
    r := len(num)-1
    
    for l+1<r {
        mid := l+(r-l)/2
        if num[mid] > target {
            r = mid
        } else {
            l = mid
        }
    }
    
    if num[l] > target {
        return l
    }
    if num[r] > target {
        return r
    }
    return r+1
}

func getRange(nums [][]int) (int, int) {
    min := math.MaxInt32
    max := math.MinInt32
    
    for _, num := range nums {
        if num[0] < min {
            min = num[0]
        }
        if num[len(num)-1] > max {
            max = num[len(num)-1]
        }
    }
    return min, max
}

File: lintcode/python/0931_median_of_k_sorted_arrays.py
#!/usr/bin/python -t

# divid and conqure and binary search
# time O(klogmlogn) m is the min val to max val range in the nums
# space O(1)


from typing import (
    List,
)

import heapq

class Solution:
    """
    @param nums: the given k sorted arrays
    @return: the median of the given k sorted arrays
    """
    def find_median(self, nums: List[List[int]]) -> float:
        # write your code hereg
        if not nums:
            return 0.0

        n = sum(len(arr) for arr in nums)
        if n == 0:
            return 0.0

        if n % 2 == 1:
            return self.find_kth(nums, n // 2 + 1) * 1.0
        return (self.find_kth(nums, n // 2) + self.find_kth(nums, n // 2 + 1)) / 2.0

    # k is one-based, from 1 to n
    def find_kth(self, arrs, k):
        start, end = self.get_range(arrs)

        # find the first num that there are >= k nums <= num
        while start + 1 < end:
            mid = (start + end) // 2
            if self.get_smaller_or_equal(arrs, mid) >= k:
                end = mid
            else:
                start = mid
        if self.get_smaller_or_equal(arrs, start) >= k:
            return start
        return end

    def get_range(self, arrs):
        start = min(arr[0] for arr in arrs if len(arr))
        end = max(arr[-1] for arr in arrs if len(arr))
        return start, end

    def get_smaller_or_equal(self, arrs, val):
        count = 0
        for arr in arrs:
            count += self.get_smaller_or_equal_in_arr(arr, val)
        return count

    def get_smaller_or_equal_in_arr(self, arr, val):
        if not arr:
            return 0

        # find the first num > val
        start, end = 0, len(arr) - 1
        while start + 1 < end:
            mid = (start + end) // 2
            if arr[mid] > val:
                end = mid
            else:
                start = mid

        if arr[start] > val:
            return start
        if arr[end] > val:
            return end
        return end + 1




# heap
# time O(klogn)
# space O(k)

from typing import (
    List,
)

import heapq

class Solution:
    """
    @param nums: the given k sorted arrays
    @return: the median of the given k sorted arrays
    """
    def find_median(self, nums: List[List[int]]) -> float:
        # write your code hereg
        n = len(nums)
        if n == 0:
            return 0

        cnt = self.get_total(nums)

        left, right = self.find_mid(nums, cnt)

        if cnt % 2 == 0:
            return (left + right) / 2.0
        else:
            return right / 1.0


    def get_total(self, nums):
        cnt = 0

        for num in nums:
            cnt += len(num)

        return cnt

    def find_mid(self, nums, cnt):
        left_index = cnt // 2
        right_index = cnt//2 + 1
        index = 0
        left, right = 0, 0

        hq = []

        for i in range(len(nums)):
            if nums[i]:
                heapq.heappush(hq, (nums[i][0], i, 0))

        while hq:
            val, row, col = heapq.heappop(hq)
            index += 1
            if index == left_index:
                left = val
            if index == right_index:
                right = val
                break

            if col+1 < len(nums[row]):
                heapq.heappush(hq, (nums[row][col+1], row, col+1))

        return left, right
